src/Platforms/OMPTarget/OMPallocator.hpp

```diff
@@ -71,12 +109,20 @@ struct OMPallocator : public HostAllocator
     HostAllocator::deallocate(pt, n);
   }
 
-  T* getDevicePtr() { return device_ptr; }
-  const T* getDevicePtr() const { return device_ptr; }
+  T* getDevicePtr() { return device_ptr_; }
+  const T* getDevicePtr() const { return device_ptr_; }
+
+  T* getDevicePtr(T* host_ptr) { return getOffloadDevicePtr(allocator_host_ptr_, host_ptr); }
+  const T* getDevicePtr(T* host_ptr) const { return getOffloadDevicePtr(allocator_host_ptr_, host_ptr); }
 
 private:
   // pointee is on device.
-  T* device_ptr = nullptr;
+  T* device_ptr_                  = nullptr;
+  T* allocator_host_ptr_ = nullptr;
+
+public:
+  T* get_device_ptr() const { return device_ptr_; }
+  T* get_allocator_host_ptr() const { return allocator_host_ptr_; }
```

Comment:
The reason we have device_ptr_ is because it is not stored anywhere else.
The allocated host has been stored in the container. So really prefer not to store a second copy.

I believe there is some design issue but without some example code in a test, I don't see the full picture of the problem you are trying to solve.

Reply:
I think we need to store a copy of it to properly abstract the container level from the details of the stateful relationship between the host and device ptrs supplied by the allocator.

Reply:
Just before closing this, we don't need this for OMPAllocator because openmp maintains mapping ranges from host to device ptrs.  We do need it in the case of the general DualAllocator where there is no runtime looking out for these mappings.

Reply:
I still think it is not needed in DualAllocator except for adding assertions. Anyway, the impact has been minimized.